Article

davschne opened the pull request:
Here's a technical article I'm planning to publish on some blogging platform or other (hopefully one that accepts markdown). I'm interested in any feedback, but particularly if anything is unclear.

To see the rendered markdown (including the images) look here: https://github.com/ihmeuw/leaflet.tilelayer.glcolorscale/blob/article/article/article.md

Commit:
Add article (in progress) on the genesis of the project.

Commit:
Revise opening paragraph of article.

Commit:
More revisions to article.

Commit:
Add section to article on the basic idea of the revised approach.

Commit:
Revise article, "idea" section.

Commit:
Add to article: section on reading floats from a WebGL texture.

Commit:
Revise article text.

Commit:
Revise capitalization scheme for article headings.

Commit:
Add a paragraph to article describing pixel colorization.

Commit:
Add section to article about getting pixels from PostGIS.

Commit:
Add a bit to section on getting pixel data to the client.

Commit:
Add section (in progress) on the Leaflet integration.

Commit:
Add more to the Leaflet integration section.

Commit:
Revise article text.

Commit:
Add "Flourishes" section to article.

With placeholder for possible inclusion of animated GIFs.

Commit:
Remove article outline.

Commit:
Revise article text.

Commit:
Add final "Outcome" section to article.

Commit:
Crop image used in article.

Commit:
Add animated GIF showing pixel transitions.

Commit:
Add article byline.

Ryshackleton commented on article/article.md:
```diff
@@ -0,0 +1,153 @@
+
+## The Idea: Client-side rendering with WebGL
+
+Around this time, I began thinking about other ways to render our map tiles. One option was to create our own tile server to do the colorization. A custom server could be better suited to our specific needs, and, crucially, we'd be more able optimize its performance because we'd understand it better. Yet a server-side solution had some inherent limitations. Rendering a large number of pixels on t ...
```
I would recommend avoiding contractions "we'd" -> "we would " in the first sentence in two places.

I would also just write out Virtual Machines. I'm sure most people know what VM's are, but you never know.  I might also reword the sentence that start "With commodity hardware..." to be something like: "Because our applications are deployed on commodity hardware running containerized services in Virtual Machines, there would be hard limits on how fast the tiles could be produced."

davschne replied:
Revised this part to better communicate the hardware limitations of a server-side approach

davschne replied:
I eliminated "we'd." Since there are plenty of other contractions in the article, I'm curious if you'd propose any general principles on the use of contractions.

Ryshackleton replied:
I would suggest not using contractions in general. I like the writing style here for a technical article, but I think it will come across as a bit more professional without contractions.

Ryshackleton commented on article/article.md:
```diff
@@ -0,0 +1,153 @@
+
+Around this time, I began thinking about other ways to render our map tiles. One option was to create our own tile server to do the colorization. A custom server could be better suited to our specific needs, and, crucially, we'd be more able optimize its performance because we'd understand it better. Yet a server-side solution had some inherent limitations. Rendering a large number of pixels on t ...
+
+Having the raw floating-point pixel values available clientside would have another advantage as well - allowing the user quick access to the actual pixel-level estimates in the visualization. While humans generally have a pretty keen color sense, the result of mentally translating a color from a gradient back to the value it's supposed to represent seemed pretty imprecise. Our visualization alrea ...
```
"clientside" should probably be two words.

This sentence is fairly obvious, distracting, and doesn't add much to what's interesting about the technical solution you came up with: "While humans generally have a pretty keen color sense, the result of mentally translating a color from a gradient back to the value it's supposed to represent seemed pretty imprecise." I would just say that the users need to see the precise values that the pixels represented in a tooltip, and then explain why your solution is an improvement over the backend request.

davschne replied:
- hyphenated "client-side"
- removed the sentence to which you objected
- revised the last sentence in the paragraph to clarify that the approach would enable showing the value in a tooltip on hover

Ryshackleton commented on article/article.md:
```diff
@@ -0,0 +1,153 @@
+
+## Problem 1: Reading floats from a texture in a WebGL shader
+
+The first and most important hurdle I foresaw with this new approach was a limitation of WebGL 1.0: the lack of support for floating-point textures. A "texture," in graphics parlance, is essentially an image you pass to the GPU. Once a texture is loaded, you can read any pixel or pixels you like from it in your shader, which is a small user-defined program that runs on the GPU. WebGL 1.0, by far  ...
```
"...you can read any pixel or pixels you like from it in your shader", don't use "it" here: "you can read any pixel or pixels you like from the texture in your shader...".

davschne replied:
revised the wording here; let me know if it's clearer

Ryshackleton commented on article/article.md:
```diff
@@ -0,0 +1,153 @@
+
+Of these choices, JPEG and PNG were unacceptable, because as implemented by PostGIS, they don't support floating-point pixels. JPEG is also a lossy format, and we didn't want to lose any data. GeoTIFF would work, because it's lossless and supports floating-point pixels, but the format is rather complicated, and I suspected it wouldn't be trivial to parse. Ultimately I opted for WKB, because its d ...
+
+With the floating-point pixel data extracted from PostGIS, I just needed a way to get it to the client. There were many possibilities for the interchange format, but I chose PNG, primarily because it seemed to be the most common format for raster data on map tile servers. Ordinarily, PNGs treat pixels as RGB or RGBA color channels, but I took inspiration from a common approach to creating [Digita ...
```
I'm a little lost in here about why you needed to extract from PostGIS as WKB, then convert to PNG. I'm not really sure how to address this. Maybe another image would help?  Some sort of simple sketch representing what's stored in each of the RGBA channels might help as well.

davschne replied:
You mean, why the two steps, instead of (a) extracting straight to PNG or (b) passing just the raw pixel data to the client?

Ryshackleton commented on article/article.md:
```diff
@@ -0,0 +1,153 @@
+
+## Flourishes
+
+By now, the main work was complete, but there were a few more nice-to-have features I was keen to add. First, I wanted to make the floating-point value of the pixel under the cursor available to user-defined mouse events. Since the parent class, `GridLayer`, maintains a cache of the tiles visible onscreen, doing so was pretty easy. I could use the cursor's screen position to determine (1) which t ...
```
I would add a GIF of your tooltips too. Maybe even with a network tab showing no network requests happening!

davschne replied:
Yeah, I really wanted to do that; just couldn't get it to work with the program (Giphy) I used to create the other animated GIF.

davschne replied:
Incorporated the screen capture you made - thanks very much!

Ryshackleton commented on article/article.md:
```diff
@@ -0,0 +1,153 @@
+
+![under-5 mortality 2000-2015 animation](./under5_mortality_animated.gif)
+
+To implement these transitions, I'd need to modify the WebGL renderer. Specifically, two textures would need to be loaded during transitions - one containing the old tiles and one containing the new. The shader code would need to compute the values of pixels in both the old and new tiles and interpolate between them over time. There were multiple cases to account for, because it was possible to ( ...
```
"Since this wasn't possible..." - Maybe make more clear that the third option is the one that isn't possible in LBD: in other words: what does "this" refer to?

davschne replied:
Changed the language in an attempt to clarify.

Ryshackleton commented on article/article.md:
```diff
@@ -0,0 +1,153 @@
+
+## Outcome
+
+The new approach to fetching and rendering map tiles in the LBD application yielded encouraging results in terms of performance and usability. Tile requests now complete much faster - generally in less than a second, compared with 3+ seconds in the prior implementation. This makes using the tool in geospatial (i.e. raster) mode a smoother experience, eliminating the frustration of having to wait  ...
```
Ugh, don't start sentences with "And". "Finally, animated per-pixel transitions..."

Ryshackleton reviewed (commented):
Hi David,

This is a great summary of quite a bit of really impressive work! I'm impressed by how well you've documented it. I had a few comments below.  Let me know if you have questions.

Ryan

Commit:
Clarify wording with respect to demonstrative.

Commit:
Clarify wording in discussion of WebGL textures.

Commit:
Don't start a sentence with "and."

Commit:
Revise text based on Ryan's suggestions.

- we'd -> we would
- clientside -> client-side
- remove unnecessary sentence
- provide clarification on the limitations of server-side rendering with respect to GPU processing

Commit:
Add animated GIF showing pixel hover feature.

Commit:
Add citations to images used in article.

Commit:
Insert line breaks in article code examples to prevent overflow.

Commit:
Change language annotation in article code block.

JavaScript, not TypeScript - though technically it's valid TypeScript (as is any JavaScript code), it includes no TypeScript features

Commit:
Eliminate stylistically incongruous sentence.
